controller/labeling_access_link/manager.py

```diff
@@ -0,0 +1,155 @@
+    information_source_item = information_source.get(project_id, heuristic_id)
+    if not information_source_item:
+        raise ValueError(f"Invalid Heuristic ID: {heuristic_id}")
+    if information_source_item.type != enums.InformationSourceType.CROWD_LABELER.value:
+        raise ValueError(f"Heuristic {heuristic_id} not meant for crowd labeling")
+
+    link = f"/projects/{project_id}/labeling/{heuristic_id}?pos=1&type={enums.LinkTypes.HEURISTIC.value}"
+    return labeling_access_link.create(
+        project_id,
+        link,
+        enums.LinkTypes.HEURISTIC,
+        created_by,
+        heuristic_id=heuristic_id,
+        with_commit=True,
+    )
+
+
+def generate_data_slice_access_link(
+    project_id: str, created_by: str, slice_id: str
+) -> LabelingAccessLink:
+    link = f"/projects/{project_id}/labeling/{slice_id}?pos=1&type={enums.LinkTypes.DATA_SLICE.value}"
+    return labeling_access_link.create(
+        project_id,
+        link,
+        enums.LinkTypes.DATA_SLICE,
+        created_by,
+        data_slice_id=slice_id,
+        with_commit=True,
+    )
+
+
+def create(
+    project_id: str,
+    link: str,
+    link_type: str,
+    created_by: str,
+    data_slice_id: Optional[str] = None,
+    heuristic_id: Optional[str] = None,
+) -> LabelingAccessLink:
+
+    try:
+        link_type_parsed = enums.LinkTypes[link_type.upper()]
+    except KeyError:
+        raise ValueError(f"Invalid LinkTypes: {link_type}")
+    return labeling_access_link.create(
+        project_id,
+        link,
+        link_type_parsed,
+        created_by,
+        data_slice_id,
+        heuristic_id,
+        None,
+        True,
+    )
+
+
+def remove(link_id: str) -> str:
+    link = labeling_access_link.get(link_id)
+    to_return = str(link.heuristic_id) if link.heuristic_id else str(link.data_slice_id)
+    labeling_access_link.remove(link_id, True)
+    return to_return
+
+
+def change_user_access_to_link_lock(link_id: str, lock_state: bool) -> str:
+    link = labeling_access_link.change_by_id(
+        link_id, {"is_locked": lock_state}, with_commit=True
+    )
+    if link:
+        return __get_type_id(link)
+
+
+def set_changed_for(project_id: str, type: enums.LinkTypes, id: str) -> None:
+    to_change = labeling_access_link.get_all_by_type_and_external_id(
+        project_id, type, id
+    )
+    for link in to_change:
+        labeling_access_link.change(link, changes=None, with_commit=False)
+        notification.send_organization_update(
+            project_id,
+            f"access_link_changed:{str(link.id)}:{__get_type_id(link)}:{link.is_locked}",
+        )
+    general.commit()
+
+
+def __get_type_id(link: LabelingAccessLink) -> str:
+    return str(
+        link.heuristic_id
+        if link.link_type == enums.LinkTypes.HEURISTIC.value
+        else link.data_slice_id
+    )
+
+
+def check_link_locked(project_id: str, link_route: str) -> bool:
+
+    if link_route.find("00000000-0000-0000-0000-000000000000") > -1:
```

Comment:
as this is listed multiple times, I'd suggest to put this into a constant/enum :)

Reply:
changed